src/Language/AST/NamedTypeNode.php

```diff
@@ -9,6 +9,6 @@ class NamedTypeNode extends Node implements TypeNode
     /** @var string */
     public $kind = NodeKind::NAMED_TYPE;
 
-    /** @var NameNode|string */
+    /** @var NameNode */
```

Comment:
I'm not sure I understand what you're uo to, here. For the moment, `NamedTypeNode::$name` is legitimately both a `NameNode` and a `string` at different times. That may or may not be what is supposed to happen, but if it's not we should figure it out and fix it for real, and if it is, I'm not sure manipulating phpstan to pretend it's not happening is making things any more clear.

Reply:
The only time where it is a `string` is within `Printer`. As a user of this library, it is always a `NameNode`.

I think we do not need to complicate usage because we misuse this class in a single part of the library.

Reply:
> That single usage is out of order

As I'm learning in my own branch, all nodes are actually being subjected to the same misuse, not just `NamedTypeNode`. Let me continue my study today, and I'll be in a better position to advise once I understand the problem better.

Reply:
You are correct that other places are also affected. This is the only one where no string coercion happens, so PHPStan rightfully complains.

Reply:
The problem with typing AST is that it is an arbitrary data structure that could be transformed into anything using a Visitor. 

In other words, this `string` typing (and actually ANY other type for ANY node) is only possible during this call within an intermediate data structure:

```php
$literallyAnything = visit($ast, $visitor);
```

Nowhere during this call we modify the original `$ast`. There is a transient copy on each step that is transformed. 

We shouldn't add all possible types to AST nodes because they can be literally anything. The only reason why we have `string` here is that we have only one utility (Printer) which transforms AST to string.

But `Visitor` supports ANY kind of transformation. It can also be `int` or `array`, so if we want "pure" types we must literally type it as `mixed` or remove the Visitor.

Which is a point where typings start bringing more harm than good? I like them as a convenience tool to help developers but not as the final truth dictating how to structure the program. Otherwise, I would have picked Java or another strictly typed language.

Reply:
I guess we could remove argument type hints from the printer Visitor which would indicate that this is not really an AST node instance but an intermediate structure.

Reply:
Removed, added a PHPDocs that explains why for future reference.

Reply:
> The problem with typing AST is that it is an arbitrary data structure that could be transformed into anything using a Visitor.

Yes, you can do anything you want in the visitor pattern, that is the point of the visitor pattern. But in our particular implementation you are calling `clone` on each node, and in PHP, that clone has the same type as the original. You then violate the types on the clone. I understand that this is a port of the javascript algorithm, but the javascript algorithm operates on javascript objects, not class instances.

> Nowhere during this call we modify the original $ast. There is a transient copy on each step that is transformed.

Again, you modify a clone, which like it or not, in PHP has the same type as the original. True, the original is not touched, but no one is accusing you of that; the problem is that the clone is being misused, and phpstan is detecting that. 

 > We shouldn't add all possible types to AST nodes because they can be literally anything. The only reason why we have string here is that we have only one utility (Printer) which transforms AST to string.

I'm a bit late in learning about the Visitor pattern, so I'm not really the best one to lecture about this, but every single visitor pattern I've looked at (I spent last night reading blogs and sample repos, watching videos, and studying up on the wiki page) does indeed switch on type. That is exactly how the visitor pattern works; it's a method for adding functionality to a family of classes without changing the classes, rather then subclassing or adding functionality to them, and to do this it has to do a lot of type checking, 

Now, one thing to keep in mind is that we have two pieces to our printer visitor puzzle: we have the `Visitor` class, and we have our `Printer` class. The `Printer` is actually the true implementation of the visitor pattern; it considers the type of each node, and adds implements its specialized functionality (producing a string). The `Visitor` class, is just a helper that iterates over the tree structure without recursion (which is apparently not a requirement of the visitor pattern). Making a copy of the node or transforming it in any way is not a key facet of the visitor pattern, and if it happens should probably happen in the actual visitor pattern implementation (meaning the `Printer`) and not in the helper.

> Which is a point where typings start bringing more harm than good? I like them as a convenience tool to help developers but not as the final truth dictating how to structure the program. Otherwise, I would have picked Java or another strictly typed language.

This has nothing to do with whether types are good or bad. The visitor pattern switches on type. That's how it works. We're okay with that, so far. Our only problem is that our visitor implementation--in this case a `Printer`--is doing its work in a graceless way. Okay, so it wants to take a tree of AST nodes and convert it to a string. Fine, this is something that happens all the time in compilers and language processors. The only problem here is that rather than creating a true intermediary format and storing it in visitor space (meaning the printer), there is a clone happening in the `Visitor` class (the helper) itself, and then the visitor implementation (the printer) is using that clone as raw storage to stuff strings into, without any kind of graceful handling. 

You're sort of getting away with it for now because those members are not currently typed at the native language level, but imagine what's going to happen to this pattern once we get around to supporting PHP 7.4 and start trying to type the node members. The architectural messiness of the current printer system will break down, and you will be forced to solve the problem properly.

My vote is to solve the problem properly, and it's what I'm currently attempting to do in my branch.

Reply:
> I guess we could remove argument type hints from the printer Visitor closures which would indicate that this is not really an AST node instance but an intermediate structure.

No, please stop. You are making things worse. The visitor pattern does indeed switch on type. That's what it does. Our problem is not that we are using types. Our problem is that our intermediary format is not an intermediary format--it's literally a clone of the original instance, and so it has the same types as the original. If you want to use an intermediary format you can, but _make a new format_... don't just use the original and scramble its guts.

Reply:
> but imagine what's going to happen to this pattern once we get around to supporting PHP 7.4 and start trying to type the node members

We'll live in the Java world then? I feel we are already half-way there :) Then why don't we use Java instead? Type hints in PHP are this: hints.

I've seen this several times - how adding strict types make things that used to be simple much harder and taking hours and days just to satisfy the type checker/compiler instead of writing something useful. I am not subscribing to 100% compatibility with PHPStan. We could have used this time to increase code coverage but instead, we fight with the type checker.

I still think we should use it as a tool where it makes sense but don't be fanatics with static analysis - the balance is important.

Reply:
> We'll live in the Java world then? I feel we are already half-way there :) Then why don't we use Java instead? Type hints in PHP are this: hints.

False equivalency. The problem we are dealing with has nothing to do with whether typing is good or bad, or being taken too far. We have the types already, like it or not. The fact that our current version of PHP allows us to trash the types does not mean that we have to, nor does it mean that our current algorithm is doing things correctly. It obviously has problems, because here we are.

> I've seen this several times - how adding strict types make things that used to be simple much harder and taking hours and days just to satisfy the type checker/compiler instead of writing something useful. I am not subscribing to 100% compatibility with PHPStan. We could have used this time to increase code coverage but instead, we fight with the type checker.

Please don't turn this into a holy war. We don't need this to devolve into some kind of weird meta discussion about whether types or good or bad. We have an architectural problem with our implementation of the visitor problem, and we can fix it. Just a reminder, this exact problem has been solved countless times in typed languages, so there's no reason we can't solve it here, too. Looseness of typing is obviously not a requirement to get this to work (I know you don't like my recursive solution in its current state, but it is an implementation of the visitor pattern with no mess).

> I still think we should use it as a tool where it makes sense but don't be fanatics with static analysis - the balance is important.

We don't have to make that choice, here. We can solve this problem in a cleaner way.